src/components/Trainer/Trainer.js

```diff
@@ -1,7 +1,327 @@
+		this.item = React.createRef();
+		this.traner = React.createRef();
+		this.files = React.createRef();
+		this.fileIndex = 0;
+
+		this.capture = this.capture.bind(this);
+		this.model = new Model(this.setStatus.bind(this));
+		this.getName = this.getName.bind(this);
+		this.addExample = this.addExample.bind(this);
+		this.train = this.train.bind(this);
+		this.predict = this.predict.bind(this);
+
+		this.state = {
+			learningRate: 0.0001,
+			batchSizeFraction: 0.4,
+			epochs: 200,
+			hiddenUnits: 100,
+			setSize: 200,
+			randomness: 0.1,
+			burstCount: 1,
+			status: 'Loading mobilenet...',
+			item: 'unknown',
+			busy: true
+		};
+	}
+
+	capture(src) {
+		return tf.tidy(() => {
+			return this.cropImage(tf.fromPixels(src))
+				.expandDims(0)
+				.toFloat()
+				.div(tf.scalar(127))
+				.sub(tf.scalar(1));
+		});
+	}
+
+	captureFromFile = async () => {
+		return new Promise(resolve => {
+			const img = new Image();
+			img.src = this.images[this.fileIndex];
+			img.onload = () => {
+				resolve(this.capture(img));
+			};
+		});
+	};
+
+	cropImage(img) {
+		const size = Math.min(img.shape[0], img.shape[1]);
+		const centerHeight = img.shape[0] / 2;
+		const beginHeight = centerHeight - size / 2;
+		const centerWidth = img.shape[1] / 2;
+		const beginWidth = centerWidth - size / 2;
+
+		return tf.image.resizeBilinear(
+			img.slice([beginHeight, beginWidth, 0], [size, size, 3]),
+			[224, 224]
+		);
+	}
+
+	screenshot = () => {
+		return this.webcam.current.webcam.current.getScreenshot();
+	};
+
+	addExample() {
+		this.setState({busy: true});
+		this.model.addExample(
+			this.screenshot,
+			() => this.capture(this.webcam.current.webcam.current.video),
+			this.state.item,
+			this.state.burstCount
+		);
+	}
+
+	addFromFile = () => {
+		this.setState({busy: true});
+
+		this.images = [];
+		this.fileIndex = 0;
+		let index = 0;
+		const reader = new FileReader();
+
+		reader.addEventListener('load', result => {
+			this.images.push(result.target.result);
+			if (index < this.files.current.files.length) {
+				reader.readAsDataURL(this.files.current.files[index++]);
+			} else {
+				this.model.addExample(
+					() => this.images[this.fileIndex++],
+					this.captureFromFile,
+					this.state.item,
+					this.files.current.files.length
+				);
+			}
+		});
+
+		reader.readAsDataURL(this.files.current.files[index++]);
+	};
+
+	train() {
+		this.setState({busy: true});
+		this.model.train(
+			this.state.hiddenUnits,
+			this.state.batchSizeFraction,
+			this.state.learningRate,
+			this.state.epochs,
+			this.state.setSize,
+			this.state.randomness
+		);
+	}
+
+	predict() {
+		if (this.webcam.current.webcam.current) {
+			this.setState({busy: true});
+			this.model.predict(
+				this.capture(this.webcam.current.webcam.current.video)
+			);
+		} else {
+			this.setStatus('Please connect a camera.');
+		}
+	}
+
+	getName(item) {
+		return item.name + (item.qualifier ? ` (${item.qualifier})` : '');
+	}
+
+	setStatus(status) {
```

Comment:
could this be combined into a single call to `setState()` to avoid 2 re-renders?

Reply:
setState is asynchronous for this reason, it won't cause 2 renders.

Reply:
[stackoverflow](https://stackoverflow.com/questions/48563650/does-react-keep-the-order-for-state-updates/48610973#48610973) says the updates are batched for lifecycle methods and event handlers - I'm assuming that this also applies here in a function called by an event handler?